experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts

```diff
@@ -0,0 +1,166 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils";
+import { type Value, type FieldKey, EmptyKey, TreeTypeSet, rootFieldKey } from "../../../core";
+// eslint-disable-next-line import/no-internal-modules
+import { leaf } from "../../../domains/leafDomain";
+import {
+	ContextuallyTypedNodeData,
+	TreeDataContext,
+	getPossibleTypes,
+	isFluidHandle,
+	isPrimitiveValue,
+} from "../../contextuallyTyped";
+import {
+	type CursorAdapter,
+	type CursorWithNode,
+	stackTreeNodeCursor,
+	stackTreeFieldCursor,
+} from "../../treeCursorUtils";
+import { type TreeNodeSchema } from "../../typed-schema";
+import { type ProxyNode } from "./types";
+
+/**
+ * Creates a {@link CursorAdaptor} capable of mapping proxy-domain trees to the editable tree domain.
+ *
+ * @param context - The {@link TreeEntity.context} of the parent node under which the tree is being inserted.
+ * @param typeSet - The types allowed under the parent node under which the tree is being inserted.
+ *
+ * @privateRemarks
+ * This is exported solely for testing purposes.
+ * Consumers should use {@link cursorFromProxyTreeNode} instead.
+ */
+export function createAdaptor<TNode extends ProxyNode<TreeNodeSchema, "javaScript">>(
+	context: TreeDataContext,
+	typeSet: TreeTypeSet,
+): CursorAdapter<TNode> {
+	return {
+		value: (node: TNode) => {
+			return isPrimitiveValue(node) || node === null || isFluidHandle(node)
+				? (node as Value)
+				: undefined;
+		},
+		type: (node: TNode) => {
+			if (node === undefined) {
+				throw new Error("undefined node");
+			}
+			if (node === null) {
+				return leaf.null.name;
+			}
+
+			if (isFluidHandle(node)) {
+				return leaf.handle.name;
+			}
+
+			switch (typeof node) {
+				case "number":
+					return leaf.number.name;
+				case "string":
+					return leaf.string.name;
+				case "boolean":
+					return leaf.boolean.name;
+				default: {
+					const possibleTypes = getPossibleTypes(
+						context,
+						typeSet,
+						node as ContextuallyTypedNodeData,
+					);
+					assert(
+						possibleTypes.length !== 0,
+						"data is incompatible with all types allowed by the schema",
+					);
+					assert(
+						possibleTypes.length === 1,
+						"data is compatible with more than one type allowed by the schema",
+					);
+					return possibleTypes[0];
+				}
+			}
+		},
+		keysFromNode: (node: TNode) => {
+			switch (typeof node) {
+				case "object":
+					if (node === null) {
+						return [];
+					} else if (Array.isArray(node)) {
+						return node.length === 0 ? [] : [EmptyKey];
```

Comment:
Why do we not have the EmptyKey field for empty arrays? Is this an optimization?

Reply:
Per our investigation, this appears to be:
a. Just part of the contract for lists
b. Consumers look for empty lists to signal no field keys, so returning an empty array will break some checks.

This contract seems ill specified IMO, but appears to be important.